Kubernetes-based worker pool

yashgorana opened the pull request:
## Description

Closes - https://github.com/OpenMined/Heartbeat/issues/894

This PR introduce Kubernetes-based worker pool
- Updates UID class to create a re-producible UUIDv4 by string
- `StatefulSet` based worker pools which gives us indexed pod name, ordered deployment+scaling, stable network identifiers.
- Updated default worker pool to spawn the same image as backend (without having to build+push a new one)


## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

yashgorana committed:
[types] uid from seed + always generate v4

yashgorana committed:
[worker] add kubernetes-based worker pool

yashgorana committed:
[k8s] fix PodStatus constructor

yashgorana committed:
[k8s] fix eager kr8s initialization

rasswanth-s committed:
re-enabled k8s tests for container workloads

yashgorana committed:
[k8s] strenum is new python

yashgorana committed:
Merge pull request #8426 from OpenMined/rasswanth/reenable-tox-test-k8s

Re-enabled k8s test

shubham3121 commented on packages/syft/src/syft/service/worker/worker_service.py:
```diff
@@ -204,7 +192,17 @@ def delete(
+            return SyftSuccess(
+                message=(
+                    f"Worker id={worker.id} was removed from pool."
+                    "Kubernetes will now re-create the pod."
```
I think we should share the message indicating, pod deletion is not supported in Kubernetes, removing and recreating the pod.

yashgorana replied:
fixed

yashgorana committed:
[syft] fix md5 security error

yashgorana committed:
[k8s] update worker delete message

shubham3121 commented on packages/syft/src/syft/service/worker/utils.py:
```diff
@@ -248,6 +254,151 @@ def run_workers_in_threads(
+        pool_pods = pool_pods[start_idx:]
+
+    # create worker object
+    for pod in pool_pods:
```
pool_pods can be SyftError, need to handle that, as `map_pod_to_worker_status` and `create_kubernetes_pool` return SyftError

yashgorana replied:
great catch - will fix this

yashgorana replied:
pushed a commit

shubham3121 commented on packages/syft/src/syft/service/worker/utils.py:
```diff
@@ -259,77 +410,99 @@ def run_containers(
-    # Worker specific environment variables go here
-    ENV SYFT_WORKER="true"
-    ENV DOCKER_TAG=${SYFT_VERSION_TAG}
+        # a very bad and hacky way to keep the Stash's unique `config` requirment happy
```
Can we add a quick `NOTE: ` on why we're doing this for reference. I realise we do this because we cannot set a local registry for the default image in prod mode, but correct me if I'm wrong. Let add some details to why this is workaround for now.

yashgorana committed:
[k8s] fix syfterror propagation

shubham3121 reviewed (approved):
Tremendous work @yashgorana !!! This is superb work 💯 🦾 🔥

shubham3121 commented:
@rasswanth-s @yashgorana Quick last thing, we need to add `container_workload` to `pytest-modules` in  `pr-tests-stack-k8s.yml`

shubham3121 committed:
Merge branch 'dev' into yash/k8s-worker

yashgorana committed:
[k8s] bad bug fix

yashgorana committed:
[k8s] return kaniko exit code in exception

yashgorana committed:
[k8s] bunch of fixes

yashgorana committed:
[k8s] fix image delete API on kubernetes

yashgorana committed:
[notebooks] update notebook for k8s runner

yashgorana committed:
[notebooks] valid pool names only

rasswanth-s committed:
Merge branch 'dev' into yash/k8s-worker

rasswanth-s committed:
Merge branch 'dev' into yash/k8s-worker

rasswanth-s committed:
separated k8s notebooks testing for container workloads

rasswanth-s committed:
revert 10-container-images back to dev branch's notebook

rasswanth-s committed:
with updated dev branch 10-container-images notebook

rasswanth-s committed:
temporarily comment out container workload stack integration tests

rasswanth-s committed:
added provision to  provide registry uid to create_image_and_pool_request

rasswanth-s committed:
updated error handling at CreateCustomImageChange

rasswanth-s committed:
added a check to prevent Image Building when deploying with K8s for Container Workloads

shubham3121 committed:
add method to get orchestration type
raise error in image deletion in case of k8s
fix pass tag directly to CreateCustomImageChange post tag validation

rasswanth-s committed:
added sleep to integration tests to wait for worker pool deletion

rasswanth-s committed:
modified WorkerOrchestrationTypeClass to Enum

rasswanth-s committed:
added selective push for kubernetes for create_image_and_pool_request

rasswanth-s committed:
updated container notebooks images with create_image_and_pool_request

rasswanth-s committed:
added step github action to add k3d-registry to /etc/hosts in github runners

rasswanth-s committed:
Added python command to github yaml file

rasswanth-s committed:
trying removing images cleanup in tox

rasswanth-s committed:
modified k3d registry port in container tests

rasswanth-s committed:
updated job retrieval in notebook

Commit:
fix job refresh in k8s notebook

Commit:
fix filter by user cod id instead of job id in k8s notebook

shubham3121 committed:
remove wait for consumer state update in k8s notebook

rasswanth-s reviewed (approved):
Awesome Work @yashgorana

yashgorana committed:
[notebooks] add some additional checks for debugging

yashgorana committed:
[ci] dump cluster info for debugging

yashgorana committed:
[ci] add pod logs

yashgorana committed:
[ci] add missing assert

yashgorana committed:
[ci] collect pr stack k8s logs

yashgorana committed:
[ci] fix k3d command not found + update devspace

rasswanth-s committed:
re-add cache removal

rasswanth-s committed:
Merge branch 'dev' into yash/k8s-worker
